### PR TITLE
[PayID] Add Skeleton of PayID Client

### PR DESCRIPTION
**Pull request by keefertaylor**
## High Level Overview of Change

Add skeleton of PayID client. 

JS: https://github.com/xpring-eng/Xpring-JS/pull/208
Swift: https://github.com/xpring-eng/XpringKit/pull/122
Java: Watch this space

### Context of Change

This change defines a client with a public API that will work with the initial payID service. All methods are unimplemented and throw `unimplemented` errors. 

Note I'm specifically not mentioning this in the CHANGELOG because it's really not ready to be used in any meaningful way and I don't want to call attention to it, yet. 

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Before / After

A new class is introduced. 

## Test Plan

Basic unit tests skeleton included.

### CODE REVIEW

File: XpringKit/PayIDClient.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  /// - Note: The input address to this method must be in X-Address format.
+  ///
+  /// - Parameters:
+  ///   - payID The payID to update.
+  ///   - xrpAddress The new XRP address to associate with the payID.
+  /// - Returns: A boolean indicating success of the operation.
+  public func update(payID: String, to xrpAddress: String) -> Result<Bool, PayIDClientError> {
+    // TODO(keefertaylor): Implement writes from backend service.
+    return .failure(.unimplemented)
+  }
```
Comment (hbergren) [inline XpringKit/PayIDClient.swift]: I think this is fine, but keep in mind that a single payment pointer can resolve to multiple addresses.

For example, you could have a Mainnet, Testnet, and Devnet XRP address, along with a BTC address all associated with a single payment pointer.

File: XpringKit/PayIDClient.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  ///   - remoteURL The remote URL of the service this class will communicate with.
+  ///   - authorizationToken An optional token to authorize write requests. Defaults to undefined.
+  public init(remoteURL: URL, authorizationToken: String? = nil) {
+    self.remoteURL = remoteURL
+    self.authorizationToken = authorizationToken
+  }
+
+  /// Retrieve the XRP Address authorized with a PayID.
+  ///
+  /// - Note: The returned value will always be in an X-Address format.
```
[the remark this answers is not in the document]
Comment (dpurdy) [reply XpringKit/PayIDClient.swift]: This was an open issue from the some other place. I would prefer that we get our canonical XRP address as the default here so it composes with the emerging XRP community support for this format, but could be convinced I am wrong.

File: XpringKit/PayIDClient.swift (hunk 1 above)
Comment (dpurdy) [reply XpringKit/PayIDClient.swift]: This is a great point for us to discuss.

1: only return one address per url per content-type?

example:
GET: doug.purdy.im CT: application/xrp -> mainnet xaddress.
GET: test.doug.purdy.im CT: application/xrp -> testnet xaddress.
GET: dev.doug.purdy.im CT: application/xrp -> devnet xaddress.

2: return N addresses per url per content-type?

GET: doug.purdy.im CT: application/xrp -> mainnet | testnet | devnet xaddresses

in any case, we need to support at else (3) three content types per PayID URL in March?

application/xrp+json
application/spsp+json
text/html (think is is @nhartner idea that I like @sappenin)

File: XpringKit/PayIDClient.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (keefertaylor) [reply XpringKit/PayIDClient.swift]: Even if this isn't true in the PayID service, I could / should convert classic addresses to X-Addresses on the fly. This is basically standard across the SDK, all classes only use X-Addresses.

File: XpringKit/PayIDClient.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  /// Initialize a new unauthorized PayIDClient.
+  ///
+  /// - Parameters:
+  ///   - remoteURL The remote URL of the service this class will communicate with.
+  public convenience init(remoteURL: URL) {
+    self.init(remoteURL: remoteURL, authorizationToken: nil)
+  }
+
+  /// Initialize a new unauthorized PayIDClient.
```
[the remark this answers is not in the document]
Comment (keefertaylor) [reply XpringKit/PayIDClient.swift]: bad paste from above; fixed.

File: XpringKit/PayIDClient.swift (hunk 1 above)
Comment (keefertaylor) [reply XpringKit/PayIDClient.swift]: My understanding here is that for PayID 1.0 (March 25 timeframe) we *only* support XRP. 

In the future, we'll support multiple currency types, and that would probably necessitate an upgrade of the SDK. My strawman proposal would be that the PayID server would deprecate the endpoint that serves XRP only addresses, and implement a new endpoint which deals with multi currency. The SDK would then deprecate this API which now connects to a deprecated endpoint and implement a new API / endpoint. 

My mental model was closer to Doug's (1). There's a testnet payID service at test.payid.xpring.io which returns testnet XRP addresses. There's payid.xpring.io which returns only mainnet XRP addresses.

----

@dpurdy I'm not sure what you mean:
application/xrp+json: This returns an XRP address
application/spsp+json: I don't know what this means? You're returning an ILP Payment pointer from a PayID? 
text/html: Similarly not sure. 

I think this API should give you either:
1) a string encoded address
2) a tuple of (a) string encoded address and (b) an enum saying what type of address it is (BTC, XRP, USD)
3) an array of the tuples in (2)

Based on my understanding of requirements for March, I went with 1. LMK if I'm misunderstanding.

File: XpringKit/PayIDClient.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (austin-king) [reply XpringKit/PayIDClient.swift]: I'm open to whatever @dpurdy and @keefertaylor think is best

File: XpringKit/PayIDClient.swift (hunk 1 above)
Comment (austin-king) [reply XpringKit/PayIDClient.swift]: application/spsp4+json returns an ILP address and shared secret, not another payment pointer. A PayID that passes in spsp4 is an ILP payment pointer. 

Yes, we've acknowledged we need to specify network alongside currency. Current thinking is that we include currency in one header and network in another.

**Comment by keefertaylor** [comment]
Thanks for discussion. I'm closing this in favor of a new implementation which resolves a bunch of comments: https://github.com/xpring-eng/Xpring-JS/pull/215

**Pull request closed without merging**